python/hail/expr/functions.py

```diff
@@ -303,6 +303,8 @@ def bind(expr, f):
 
     f_input = construct_expr(Reference(uid), expr._type, expr._indices, expr._aggregations, expr._joins, expr._refs)
     lambda_result = to_expr(f(f_input))
+    if not lambda_result._aggregations.empty():
+        raise ExpressionException("aggregators may not be used inside lambda functions")
```

Comment:
I think the issue is that `f_input` should throw an error if it's placed inside of an aggregation. This should still be OK:

```python
t.annotate(foo = hl.bind(1, lambda x: x + agg.sum(t.VAL)))
```

Maybe names need a "binding context". You can only reference names that are bound _inside_ the aggregator, like this:

```python
t.annotate(foo = agg.mean(hl.bind(... big computationally expensive expression here ...,
                                  lambda x: x * x / t.VAL)))
```

Reply:
ok, will see if I can rework this.